bindings/web/diffie_hellman.cc

```diff
@@ -0,0 +1,177 @@
+#include <emscripten.h>
+#include <emscripten/bind.h>
+#include <emscripten/val.h>
+
+#include <sstream>
+#include <iostream>
+#include <iomanip>
+
+#include <openssl/evp.h>
+#include <openssl/ec.h>
+#include <openssl/pem.h>
+#include <openssl/rand.h>
+#include <openssl/err.h>
+#include <openssl/sha.h>
+
+using namespace emscripten;
+
+#define ERROR_NULL(VAR, FUN)                                       \
+    if(_return.empty())                                            \
+    {                                                              \
+        VAR = FUN;                                                 \
+        if(VAR == NULL)                                            \
+            _return = "error-openssl:" + std::to_string(__LINE__); \
+    }
+
+#define ERRORC(FUN)                                                    \
+    {                                                                  \
+        if(_return.empty())                                            \
+        {                                                              \
+            size_t a = (size_t)FUN;                                    \
+            if(a != 1)                                                 \
+                _return = "error-openssl:" + std::to_string(__LINE__); \
+        }                                                              \
+    }
+
+class DiffieHellman
+{
+  private:
+    
+    EVP_PKEY* _private_key;
+
+  public:
+
+    DiffieHellman() : _private_key(NULL)
+    {}
+
+    void free()
+    {
+        EVP_PKEY_free(this->_private_key);
+    }
+
+    std::string initialize()
+    {
+        std::string _return;
+
+        EVP_PKEY_CTX* parameters_context = NULL;
+        EVP_PKEY_CTX* key_generation_context = NULL;
+        EVP_PKEY* params = NULL;
+        this->_private_key = NULL;
+
+        ERROR_NULL(parameters_context, EVP_PKEY_CTX_new_id(EVP_PKEY_EC, nullptr));
+
+        ERRORC(EVP_PKEY_paramgen_init(parameters_context));
+
+        ERRORC(EVP_PKEY_CTX_set_ec_paramgen_curve_nid(parameters_context, NID_X9_62_prime256v1));
+        ERRORC(EVP_PKEY_CTX_set_ec_param_enc(parameters_context, OPENSSL_EC_NAMED_CURVE));
+        ERRORC(EVP_PKEY_paramgen(parameters_context, &params));
+
+        ERROR_NULL(key_generation_context, EVP_PKEY_CTX_new(params, nullptr));
+        ERRORC(EVP_PKEY_keygen_init(key_generation_context));
+        ERRORC(EVP_PKEY_keygen(key_generation_context, &this->_private_key));
+
+        EVP_PKEY_CTX_free(parameters_context);
+        EVP_PKEY_CTX_free(key_generation_context);
+        EVP_PKEY_free(params);
+
+        return _return;
+    }
+
+    std::string get_public_key()
+    {
+        if(this->_private_key == NULL) return "error: not initialized.";
+
+        std::string public_key_vec;
+        std::string	_return;
+
+        {
```

Comment:
Is this block needed?

Reply:
I used this block for readability, to `isolate` reading from input variable to the rest of the operations. Also variables used to read from input have more delimited life scope.